### PR TITLE
Add job client

### PR DESCRIPTION
**Pull request by zolvarga**
# Thank you for helping us improve the Azure IoT Python SDK!

## Need Support

- **Have a feature request for SDKs?** Please post it on [User Voice](https://feedback.azure.com/forums/321918-azure-iot) to help us prioritize
- **Have a technical question?** Ask on [Stack Overflow with tag "azure-iot-hub"](https://stackoverflow.com/questions/tagged/azure-iot-hub)
- **Need Support?** Every customer with an active Azure subscription has access to [support](https://docs.microsoft.com/en-us/azure/azure-supportability/how-to-create-azure-support-request) with guaranteed response time.  Consider submitting a ticket and get assistance from Microsoft support team
- **Found a bug?** Please help us fix it by thoroughly documenting it and filing an issue on GitHub (See below).

Here's a little checklist of things that will help it make its way to the repository: Note that you don't have to check all the boxes, we can help you with that.
This being said, the more you do, the quicker it'll go through our gated build!
-->

## Checklist

- [ ] I have read the [contribution guidelines](https://github.com/Azure/azure-iot-sdk-python/.github/CONTRIBUTING.md).
- [ ] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- If this is a modification that impacts the behavior of a public API
  - [ ] I edited the corresponding document in the `devdoc` folder and added or modified requirements.

## Description of the problem

Please be as precise as possible: what issue you experienced, how often...

## Description of the solution

How you solved the issue and the other things you considered and maybe rejected

### CODE REVIEW

File: azure-iot-hub/azure/iot/hub/iothub_job_manager.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        :rtype: :class:`azure.iot.hub.IoTHubJobManager`
+        """
+
+        self.auth = ConnectionStringAuthentication(connection_string)
+        self.protocol = protocol_client(self.auth, "https://" + self.auth["HostName"])
+
+    def create_import_export_job(self, job_properties):
+        """Creates a new import/export job on an IoT hub.
+
+        :param job_properties job_properties: Specifies the job specification.
```
Comment (cartertinney) [inline azure-iot-hub/azure/iot/hub/iothub_job_manager.py:35]: >job_properties [](start = 15, length = 14)

This is not a type

File: azure-iot-hub/azure/iot/hub/iothub_job_manager.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+        :raises: `HttpOperationError<msrest.exceptions.HttpOperationError>`
+            if the HTTP response status is not in [200].
+
+        :returns: JobProperties object containing the created job.
```
Comment (cartertinney) [inline azure-iot-hub/azure/iot/hub/iothub_job_manager.py:40]: >:returns: JobProperties object containing the created job. [](start = 8, length = 58)

You also will want to have a line below this like

:rtype: :class:`azure.iot.hub.models.JobProperties`

File: azure-iot-hub/azure/iot/hub/iothub_job_manager.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        """
```
Comment (cartertinney) [inline azure-iot-hub/azure/iot/hub/iothub_job_manager.py:41]: Is this a new JobProperties object? Or the same instance that was passed in?

File: azure-iot-hub/azure/iot/hub/iothub_job_manager.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    def cancel_import_export_job(self, job_id):
+        """Cancels an import/export job on an IoT hub.
+
+        :param str job_id: The ID of the job.
+
+        :raises: `HttpOperationError<msrest.exceptions.HttpOperationError>`
+            if the HTTP response status is not in [200].
+
+        :returns: Object.
```
Comment (cartertinney) [inline azure-iot-hub/azure/iot/hub/iothub_job_manager.py:74]: >Object. [](start = 18, length = 7)

?

File: azure-iot-hub/tests/test_iothub_http_runtime_manager.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        skn=fake_shared_access_key_name,
+        sk=fake_shared_access_key,
+    )
+    iothub_http_runtime_manager = IoTHubHttpRuntimeManager(connection_string)
+    return iothub_http_runtime_manager
+
+
+@pytest.mark.describe("IoTHubHttpRuntimeManager - .receive_feedback_notification()")
+class TestReceiveFeedbackNotification(object):
+    @pytest.mark.it("Receive feedback notification")
```
Comment (cartertinney) [inline azure-iot-hub/tests/test_iothub_http_runtime_manager.py:45]: >Receive feedback notification [](start = 21, length = 29)

Here and everywhere, more descriptive test descriptions, in the grammatical format as if the sentence has started with "It"

e.g. "Uses protocol layer HTTP runtime to receive feedback notifications"

File: azure-iot-hub/tests/test_iothub_http_runtime_manager.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    def test_receive_feedback_notification(
+        self, mocker, mock_http_runtime_operations, iothub_http_runtime_manager
+    ):
+        iothub_http_runtime_manager.receive_feedback_notification()
+        assert mock_http_runtime_operations.receive_feedback_notification.call_count == 1
+        assert mock_http_runtime_operations.receive_feedback_notification.call_args == mocker.call()
+
```
Comment (cartertinney) [inline azure-iot-hub/tests/test_iothub_http_runtime_manager.py:52]: Here and everywhere, there is no test of returned values, only calls made.

File: azure-iot-hub/tests/test_iothub_http_runtime_manager.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        hostname=fake_hostname,
+        device_id=fake_device_id,
+        skn=fake_shared_access_key_name,
+        sk=fake_shared_access_key,
+    )
+    iothub_http_runtime_manager = IoTHubHttpRuntimeManager(connection_string)
+    return iothub_http_runtime_manager
+
+
+@pytest.mark.describe("IoTHubHttpRuntimeManager - .receive_feedback_notification()")
```
Comment (cartertinney) [inline azure-iot-hub/tests/test_iothub_http_runtime_manager.py:43]: >IoTHubHttpRuntimeManager [](start = 23, length = 24)

Missing instantiation test(s)

File: azure-iot-hub/tests/test_job_manager.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
+# --------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
```
Comment (cartertinney) [inline azure-iot-hub/tests/test_job_manager.py:6]: All comments for the HTTP runtime manager tests apply to this file as well

File: azure-iot-hub/azure/iot/hub/iothub_job_manager.py (hunk 3 above)
Comment (zolvarga) [reply azure-iot-hub/azure/iot/hub/iothub_job_manager.py:41]: A new one, what has been created

---
In reply to: [405794291](https://github.com/Azure/azure-iot-sdk-python/pull/532#discussion_r405794291) [](ancestors = 405794291)

File: azure-iot-hub/azure/iot/hub/iothub_job_manager.py (hunk 4 above)
Comment (zolvarga) [reply azure-iot-hub/azure/iot/hub/iothub_job_manager.py:74]: That is the type from the generated code.

---
In reply to: [405794483](https://github.com/Azure/azure-iot-sdk-python/pull/532#discussion_r405794483) [](ancestors = 405794483)

File: azure-iot-hub/tests/test_job_manager.py [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    )
+    iothub_job_manager = IoTHubJobManager(connection_string)
+    return iothub_job_manager
+
+
+@pytest.mark.describe("IoTHubJobManager")
+class TestJobManager(object):
+    @pytest.mark.it("Instantiates with an empty connection string")
+    def test_instantiates_with_empty_connection_string(self):
+        with pytest.raises(Exception):
```
Comment (cartertinney) [inline azure-iot-hub/tests/test_job_manager.py:51]: >Exception [](start = 27, length = 9)

Never do this in pytests. Capture the SPECIFIC exception being raised. Capturing a generic Exception will catch anything, including unrelated errors, and actual failures in testing. We've had this problem a lot in device client, where many bugs were missed due to this exact mistake. Here and everywhere.

File: azure-iot-hub/tests/test_job_manager.py [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+        skn=fake_shared_access_key_name,
+        sk=fake_shared_access_key,
+    )
+    iothub_job_manager = IoTHubJobManager(connection_string)
+    return iothub_job_manager
+
+
+@pytest.mark.describe("IoTHubJobManager")
+class TestJobManager(object):
+    @pytest.mark.it("Instantiates with an empty connection string")
```
Comment (cartertinney) [inline azure-iot-hub/tests/test_job_manager.py:49]: >Instantiates with an empty connection string [](start = 21, length = 44)

Seems like what it does is actually "Raises a <Exception Name> exception when instantiated with an empty connection string".

Here and everywhere.

File: azure-iot-hub/tests/test_job_manager.py [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+        hostname=fake_hostname,
+        device_id=fake_device_id,
+        skn=fake_shared_access_key_name,
+        sk=fake_shared_access_key,
+    )
+    iothub_job_manager = IoTHubJobManager(connection_string)
+    return iothub_job_manager
+
+
+@pytest.mark.describe("IoTHubJobManager")
```
Comment (cartertinney) [inline azure-iot-hub/tests/test_job_manager.py:47]: >(" [](start = 21, length = 2)

There's also no tests here for the happy path - i.e. testing that the attributes of the instance of the class are set as expected

File: azure-iot-hub/tests/test_job_manager.py [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+class TestJobManager(object):
+    @pytest.mark.it("Instantiates with an empty connection string")
+    def test_instantiates_with_empty_connection_string(self):
+        with pytest.raises(Exception):
+            IoTHubJobManager("")
+
+    @pytest.mark.it("Instantiates with an connection string without HostName")
+    def test_instantiates_with_connection_string_no_host_name(self):
```
Comment (cartertinney) [inline azure-iot-hub/tests/test_job_manager.py:55]: You could parametrize all these "bad path" tests that are expected to raise the same error into a single test

File: azure-iot-hub/tests/test_job_manager.py [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+        device_id=fake_device_id,
+        skn=fake_shared_access_key_name,
+        sk=fake_shared_access_key,
+    )
+    iothub_job_manager = IoTHubJobManager(connection_string)
+    return iothub_job_manager
+
+
+@pytest.mark.describe("IoTHubJobManager")
+class TestJobManager(object):
```
Comment (cartertinney) [inline azure-iot-hub/tests/test_job_manager.py:48]: "IoTHubJobManager - Instantiation"

**Pull request merged**